### DISCUSSION TITLE
Error when adding add_special_case

### DISCUSSION THREAD

**Discussion by ghost** [explosion/spaCy#12447] [Help: Other Questions] [answered]
## Info about spaCy

- **spaCy version:** 3.5.1
- **Platform:** Windows-10-10.0.19045-SP0
- **Python version:** 3.9.6
- **Pipelines:** ru_core_news_md (3.5.0), ru_core_news_sm (3.5.0)

Greetings! I'm getting error when i'm trying add some special cases.
Error message:

C:\Users\spos\PycharmProjects\parseforradio\venv\Scripts\python.exe C:/Users/spos/PycharmProjects/parseforradio/stress.py
Traceback (most recent call last):
  File "C:\Users\spos\PycharmProjects\parseforradio\stress.py", line 158, in <module>
    introduce_special_cases_from_dictionary(wordforms)
  File "C:\Users\spos\PycharmProjects\parseforradio\stress.py", line 19, in introduce_special_cases_from_dictionary
    ru_nlp.tokenizer.add_special_case(word, [{"ORTH": dictionary[word][0]["accentuated"]}])
  File "spacy\tokenizer.pyx", line 604, in spacy.tokenizer.Tokenizer.add_special_case
  File "spacy\tokenizer.pyx", line 588, in spacy.tokenizer.Tokenizer._validate_special_case
ValueError: [E997] Tokenizer special cases are not allowed to modify the text. This would map 'палочки для еды' to 'па+лочки для еды+' given token attributes '[{'ORTH': 'па+лочки для еды+'}]'.

My code:

```python
import spacy
import pickle
ru_nlp = spacy.load('ru_core_news_md')


def load():
    with open(file="lemmas.dat", mode='rb') as f:
        lemmas = pickle.loads(f.read())
    with open(file="wordforms.dat", mode='rb') as f:
        wordforms = pickle.loads(f.read())
    return lemmas, wordforms


def introduce_special_cases_from_dictionary(dictionary):
    for word in dictionary:
        if (" " in word) or ("-" in word):
            if len(dictionary[word]) == 1:
                ru_nlp.tokenizer.add_special_case(word, [{"ORTH": dictionary[word][0]["accentuated"]}])
                ru_nlp.tokenizer.add_special_case(word.capitalize(), [{"ORTH": dictionary[word][0]["accentuated"].capitalize()}])
lemmas, wordforms = load()
introduce_special_cases_from_dictionary(wordforms)
```

So, why it doesn't work for me?

**Comment by adrianeboyd** [comment, marked as answer]
The error message is explaining the problem: you can't change the text from `палочки для еды` to `па+лочки для еды+` using tokenizer special cases. The built-in `Tokenizer` does not allow  you to modify the underlying text.

A note: it is technically possible to have a custom tokenizer in `nlp.tokenizer` modify the input text when creating the `Doc`, but the built-in tokenizers in the `spacy` library do not support this because preserving the input text is one of spacy's key features. If you need to preprocess your texts, we usually recommend doing this in a separate step outside your spacy pipeline. Also, spacy v2 did unintentionally allow you to do this with tokenizer special cases, so if you're upgrading older code that might explain why you're noticing this now with v3.
